tools/tpm2_listpersistent.c

```diff
@@ -103,16 +105,30 @@ int readPublic(TSS2_SYS_CONTEXT *sapi_context, TPMI_DH_OBJECT objectHandle) {
 
     TPMI_ALG_PUBLIC type = outPublic.t.publicArea.type;
     TPMI_ALG_HASH nameAlg = outPublic.t.publicArea.nameAlg;
-    UINT32 attrs = outPublic.t.publicArea.objectAttributes.val;
+    char *attrs = tpm2_attr_util_obj_attrtostr(
+            outPublic.t.publicArea.objectAttributes);
+    bool should_free = !!attrs;
+    if (!attrs) {
+        LOG_WARN("Could not convert objectAttributes, converting to hex output");
+        char tmp[11]; /* UINT32 in hex (8) + "0x" + '\0' */
```

Comment:
Maybe it will be easier to read if you just allocate the array in attrs and copy there directly. That way you can get rid of the should_free and tmp variables and just check if attrs != NULL to free.

Reply:
I'm not sure what your saying but I have another approach to this that's simpler.